tfjs-backend-webgpu/src/conv_backprop_mm_webgpu.ts

```diff
@@ -16,107 +16,168 @@
+    this.tileInner = Math.max(
+        this.workGroupSize[0] * this.innerElementSize, this.workGroupSize[1]);
+    this.shaderKey = `conv2DDerInputMM_${this.isVec4}_${
+        this.elementsPerThread}_${this.innerElementSize}`;
   }
 
   getUserCode(): string {
-    const matMulSource =
+    const matMulSource = this.isVec4 ?
+        makeMatMulPackedVec4Source(
+            this.elementsPerThread, this.tileAOuter, this.tileBOuter,
+            this.tileInner, this.innerElementSize, !this.isChannelsLast) :
         makeMatMulPackedSource(this.elementsPerThread, this.workGroupSize);
-
-    const readASnippet = `
-    let outRow = row / uniforms.outShape[2];
-    let outCol = row % uniforms.outShape[2];
-
-    let WRow = col / (uniforms.filterDims[1] * uniforms.outBackprop[3]);
-    let WCol = col / uniforms.outBackprop[3] % uniforms.filterDims[1];
-    let xR = f32(outRow - uniforms.pads[0] + WRow) / f32(uniforms.stride[0]);
-    let xC = f32(outCol - uniforms.pads[1] + WCol) / f32(uniforms.stride[1]);
-    if (xR < 0.0 || xR >= f32(uniforms.outBackprop[1]) || fract(xR) > 0.0) {
-      return 0.0;
-    }
-    if (xC < 0.0 || xC >= f32(uniforms.outBackprop[2]) || fract(xC) > 0.0) {
-      return 0.0;
-    }
-    let coord = vec4<i32>(
-        batch,
-        i32(xR),
-        i32(xC),
-        col % uniforms.outBackprop[3]);
-    return x[getIndexFromCoords4D(coord, uniforms.xShape)];`;
-
-    const sampleA = `if (row < uniforms.dimAOuter && col < uniforms.dimInner) {
-      ${readASnippet}
-    }
-    return 0.0;`;
-
+    const elementsSize = this.isVec4 ?
+        [this.isChannelsLast ? this.innerElementSize : 4, 4, 4] :
+        [1, 1, 1];
     const userCode = `
-    fn mm_readA(row : i32, col : i32, globalId : vec3<u32>) -> f32 {
-      var batch = i32(globalId.z);
-      ${sampleA}
-    }
-
-    fn mm_readB(row : i32, col : i32, globalId : vec3<u32>) -> f32 {
-      let coordX = uniforms.filterDims.x - 1 -
-          row / (uniforms.filterDims[1] * uniforms.outBackprop[3]);
-      let coordY = uniforms.filterDims.y - 1 -
-          (row / uniforms.outBackprop[3]) % uniforms.filterDims[1];
-      if (row < uniforms.dimInner && col < uniforms.dimBOuter &&
-          coordX >= 0 && coordY >= 0) {
-        let coord = vec4<i32>(coordX, coordY, col,
-            row % uniforms.outBackprop[3]);
-        return W[getIndexFromCoords4D(coord, uniforms.wShape)];
-      }
-      return 0.0;
-    }
-
-    fn mm_write(row : i32, col : i32, valueInput : f32, globalId : vec3<u32>) {
-      if (row < uniforms.dimAOuter && col < uniforms.dimBOuter)
-      {
-      var batch = i32(globalId.z);
-      var value = valueInput;
-      let outCoord = vec4<i32>(
-          batch,
-          row / uniforms.outShape[2],
-          row % uniforms.outShape[2],
-          col);
-      result[getIndexFromCoords4D(outCoord, uniforms.outShape)] = value;
-      }
-    }
-
+    ${conv2dTransposeCommonSnippet(elementsSize[2])}
```

Comment:
`${conv2dTransposeCommonSnippet(this.isVec4? 4 : 1)}` ?

Reply:
done